### PR TITLE
Add tests for reduction with sum of squares (RSS), compute of RNSNorm, RNSNorm fused with Matmul, FindMaxAttention and Softmax

### PR DESCRIPTION
**Pull request by mikepapadim**
#### Description

This PR adds tests to test the code gen and and vailidity of the results for some common operation required in the LLM architecture. 

1. `Reduction with Sum of Squares (RSS)`: Ensures the correct functionality of reduction operations that compute the sum of squares, which is vital for various numerical algorithms.​

2. `Compute of RNSNorm`: Validates the implementation of the RNSNorm function, a normalization technique that can enhance the stability and performance of neural networks.​

3. `RNSNorm Fused with Matmul`: Tests the integration of RNSNorm directly within matrix multiplication operations, aiming to optimize performance by reducing the number of separate computational steps.​

4. `FindMaxAttention`: Assesses the functionality of the FindMaxAttention operation, which is essential in attention mechanisms commonly used in transformer models for tasks like natural language processing.​

5. `Softmax`: Verifies the correctness of the Softmax function implementation, a critical component in various machine learning models, particularly in classification tasks.

6. `MultiHeadedAttention`:  Verifies the correctness of the Mutiheaded attention, it breaks down individual tasks/kernels and uses a refernce Java implmentation very close to LLama3.java. 

7. `FFN`: Verifies the corectness layer.

8. `Fused with RoPE`: verifies corectness .

#### Backend/s tested

Mark the backends affected by this PR.

- [x] OpenCL
- [x] PTX
- [ ] SPIRV

#### OS tested

Mark the OS where this PR is tested.

- [x] Linux
- [ ] OSx
- [ ] Windows

#### Did you check on FPGAs?

If it is applicable, check your changes on FPGAs.

- [ ] Yes
- [x] No

#### How to test the new patch?

```bash

tornado-test -V uk.ac.manchester.tornado.unittests.reductions.TestReductionsFloats#testReduceSumSquares

tornado-test -V uk.ac.manchester.tornado.unittests.llm.TestRNSNormLayer

tornado-test -V uk.ac.manchester.tornado.unittests.llm.TestSoftMaxLayer

tornado-test -V uk.ac.manchester.tornado.unittests.llm.TestFindMaxAttention

tornado-test -V uk.ac.manchester.tornado.unittests.llm.TestParallelFFNLayer

tornado-test -V uk.ac.manchester.tornado.unittests.llm.TestMultiHeadAttention

tornado-test   -V uk.ac.manchester.tornado.unittests.llm.TestFusedLayer

```
----------------------------------------------------------------------------

### CODE REVIEW

File: tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/compute/LLMFusedKernelsTest.java [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+                if (localIdx < stride) {
+                    localA[localIdx] += localA[localIdx + stride];
+                }
+            }
+            if (localIdx == 0) {
+                reduce.set(groupID, localA[0]);
+            }
+        }
+
+            private static void finalSum(KernelContext context, FloatArray reduce, int size, float eps) {
```
Comment (jjfumero) [inline tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/compute/LLMFusedKernelsTest.java:131]: Fix format of the file.

File: tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/compute/LLMFusedKernelsTest.java [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+                executionPlan.withGridScheduler(gridScheduler) //
+                        .execute();
+            }
+
+
+            for (int i = 0; i < outputSeqLogits.getSize(); i++) {
+                float expected = outputSeqLogits.get(i);           // Expected value from the sequential output
+                float actual = outputLogits.get(i);               // Actual value from the RNS output
+
+//                assertEquals("Mismatch at index " + i, expected, actual, 1f); // Allow some tolerance
```
Comment (jjfumero) [inline tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/compute/LLMFusedKernelsTest.java:281]: Remove comment

**Review by jjfumero — CHANGES_REQUESTED** [review]

**Comment by jjfumero** [comment]
Include the new test in the test-suite: 
`tornado-assembly/src/bin/tornado-test`

**Comment by jjfumero** [comment]
The new test enters in an infinite loop when running with the SPIR-V backend:

```bash
tornado-test -V uk.ac.manchester.tornado.unittests.compute.LLMFusedKernelsTest
/home/juan/tornadovm/TornadoVM/bin/sdk/bin/tornado --jvm "-Xmx6g -Dtornado.recover.bailout=False -Dtornado.unittests.verbose=True "  -m  tornado.unittests/uk.ac.manchester.tornado.unittests.tools.TornadoTestRunner  --params "uk.ac.manchester.tornado.unittests.compute.LLMFusedKernelsTest"
```

The PTX and OpenCL backends run fine.

File: tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/compute/LLMFusedKernelsTest.java [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+     * <p>
+     * How to run the tests?
+     * </p>
+     *
+     * <code>
+     * tornado-test -V uk.ac.manchester.tornado.unittests.compute.LLMFusedKernelsTest
+     * </code>
+     */
+
+    public class LLMFusedKernelsTest extends TornadoTestBase {
```
Comment (stratika) [inline tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/compute/LLMFusedKernelsTest.java:69]: To keep consistency with other test classes, I would suggest to move the "Test" at the beginning of the name of the class.

File: tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/compute/LLMFusedKernelsTest.java [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+                    FloatArray out, FloatArray input, FloatArray weight, FloatArray scalingFactorBuffer,
+                    int size, float eps) {
+
+                int globalIdx = context.globalIdx;
+
+                if (globalIdx < size) {
+                    float scaledValue = weight.get(globalIdx) * (scalingFactorBuffer.get(0) * input.get(globalIdx));
+                    out.set(globalIdx, scaledValue);
+                }
+            }
```
Comment (stratika) [inline tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/compute/LLMFusedKernelsTest.java:157]: fix code formatting.

File: tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/compute/LLMFusedKernelsTest.java [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+
+        @Test
+        public void testRNSNorm() throws TornadoExecutionPlanException {
+            final int size = 2048;
```
Comment (stratika) [inline tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/compute/LLMFusedKernelsTest.java:161]: I think we should add the following, unless it is supported: 
```bash
assertNotBackend(TornadoVMBackendType.SPIRV);
```

File: tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/compute/LLMFusedKernelsTest.java [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+
+                // Perform the comparison with assertion
+                assertEquals("Mismatch at index " + i, expected, actual, 1e-6f); // Allow some tolerance
+            }
+
+        }
+
+        @Test
+        public void testRNSNormFusedWithMatMul() throws TornadoExecutionPlanException {
+            final int size = 2048;
```
Comment (stratika) [inline tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/compute/LLMFusedKernelsTest.java:219]: I think we should add the following, unless it is supported: 
```bash
assertNotBackend(TornadoVMBackendType.SPIRV);
```

File: tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/compute/LLMFusedKernelsTest.java (hunk 2 above)
Comment (stratika) [reply tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/compute/LLMFusedKernelsTest.java:281]: I guess the test should assert the expected and actual values, right?

**Review by stratika — COMMENTED** [review]
I think we should add the new `LLMFusedKernelsTest` class in the `tornado-test` in order to be run when Jenkins runs the unit-tests. In my setup, the tests pass for PTX. But, the tests in the `LLMFusedKernelsTest` class are not finishing when running with SPIR-V. I guess, that they are not supported for SPIR-V?

**Comment by jjfumero** [comment]
@mikepapadim , is this ready?

**Comment by jjfumero** [comment]
Thanks @mikepapadim. Did you check with SPIR-V. Last time it got stuck in an infinite loop.

**Comment by mikepapadim** [comment]
> Thanks @mikepapadim. Did you check with SPIR-V. Last time it got stuck in an infinite loop.

No, I need to switch machines for that

**Comment by stratika** [comment]
I confirm it seems to be something odd with SPIR-V. Let's wait and if it is not supported for any reason we can mark it as unsupported. Currently it seems like it is stuck. Note that we should test with both the OpenCL and Level Zero runtimes.

```bash
make BACKEND=spirv
tornado-test --quickPass --verbose --jvm="-Dtornado.unittests.device=0:0"
tornado-test --quickPass --verbose --jvm="-Dtornado.unittests.device=0:1"
```

**Comment by stratika** [comment]
I can help with the issue in SPIR-V. I will have a look next week.

**Comment by mikepapadim** [comment]
TODO: add all to tornado-test script

File: tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/llm/TestRMSNormLayer.java [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+
+        ImmutableTaskGraph immutableTaskGraph = taskGraph.snapshot();
+        try (TornadoExecutionPlan executionPlan = new TornadoExecutionPlan(immutableTaskGraph)) {
+            executionPlan.withGridScheduler(gridScheduler) //
+                    .execute();
+        }
+
+        for (int i = 0; i < outputSeqLogits.getSize(); i++) {
+            float expected = outputSeqLogits.get(i);           // Expected value from the sequential output
+            float actual = outputLogits.get(i);               // Actual value from the RNS output
```
Comment (stratika) [inline tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/llm/TestRMSNormLayer.java:271]: we are missing the `assertEquals` method for the expected and actual value.

File: tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/llm/TestParallelFFNLayer.java [hunk 8]
```diff
@@ -0,0 +1,420 @@
...
+                    allMatch = false;
+                }
+            }
+
+            if (allMatch) {
+                System.out.println("All values match within tolerance!");
+            } else {
+                System.out.println("Some values don't match within tolerance.");
+            }
+        }
```
Comment (stratika) [inline tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/llm/TestParallelFFNLayer.java:413]: Similar comment with previous tests. Can we replace the print and conditional checks with the invocation of the `assertEquals` and add the error fraction for the acceptable deviation due to the precision difference?

File: tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/llm/TestWeightedSum.java [hunk 9]
```diff
@@ -0,0 +1,359 @@
...
+        FloatArray valueCache = new FloatArray(seqLen * kvDim);
+        FloatArray output = new FloatArray(numHeads * headSize);
+        FloatArray expectedOutput = new FloatArray(numHeads * headSize);
+        FloatArray debugData = new FloatArray(numHeads * localSize * 5); // Additional debug information
+
+        // Initialize with controlled values
+        Random random = new Random(42);
+
+        // Print head 0, dimension 0 values for clarity
+        System.out.println("Initializing test data:");
```
Comment (stratika) [inline tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/llm/TestWeightedSum.java:121]: I would suggest to remove the printing statement.

File: tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/llm/TestWeightedSum.java [hunk 10]
```diff
@@ -0,0 +1,359 @@
...
+        for (int t = 0; t <= pos; t++) {
+            System.out.printf("Position %d: %.6f%n", t, attScores.get(0 * seqLen + t));
+        }
+
+        // Print the first few value vectors for head 0
+        System.out.println("First few values for head 0, dimension 0:");
+        for (int t = 0; t <= Math.min(pos, 3); t++) {
+            int valueOffset = loff + t * kvDim + (0 / kvMul) * headSize + 0;
+            System.out.printf("Position %d: %.6f%n", t, valueCache.get(valueOffset));
+        }
```
Comment (stratika) [inline tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/llm/TestWeightedSum.java:168]: Are this print statements valuable for when we run the unit-tests? or are they added for documentation purposes? If it is for documentation, perhaps we can add one flag that is FALSE by default, in order to reduce the logged output when we run all unit-tests. 

In my opinion, when we run the tests we should focus on PASS/FAIL results. What do you think?

**Comment by stratika** [comment]
Also, one interesting note that I discussed with @mikepapadim is that when we run the unit-tests with `make fast-tests`, this uses the `tornado-test --ea --verbose --quickPass` command. and the flag `--ea` which is used to enable assertions at the compiler level adds a delay when we run the following test:

```bash
tornado-test  --ea -V uk.ac.manchester.tornado.unittests.llm.TestParallelFFNLayer
```

**Comment by jjfumero** [comment]
For the SPIR-V I get the following failures:

```
/home/juan/repos/tornadovm/bin/sdk/bin/tornado --jvm "-Xmx6g -Dtornado.recover.bailout=False -Dtornado.unittests.verbose=True "  -m  tornado.unittests/uk.ac.manchester.tornado.unittests.tools.TornadoTestRunner  --params "uk.ac.manchester.tornado.unittests.llm.TestFindMaxAttention"
WARNING: Using incubator modules: jdk.incubator.vector

Test: class uk.ac.manchester.tornado.unittests.llm.TestFindMaxAttention
	Running test: testFindMaxAttentionScores ................  [FAILED] 
		\_[REASON] Mismatch at head 0 expected:<4.9773192> but was:<0.0>
Test ran: 1, Failed: 1, Unsupported: 0

```

And:

```
/home/juan/repos/tornadovm/bin/sdk/bin/tornado --jvm "-Xmx6g -Dtornado.recover.bailout=False -Dtornado.unittests.verbose=True "  -m  tornado.unittests/uk.ac.manchester.tornado.unittests.tools.TornadoTestRunner  --params "uk.ac.manchester.tornado.unittests.llm.TestMultiHeadAttention"
WARNING: Using incubator modules: jdk.incubator.vector

Test: class uk.ac.manchester.tornado.unittests.llm.TestMultiHeadAttention
	Running test: testMultiHeadAttentionLlamaEquivalence ................  [FAILED] 
		\_[REASON] expected:<-0.052507114> but was:<NaN>
	Running test: testAttentionScoresCalculation ................  [PASS] 
	Running test: testFindMaxAttentionScores ................  [FAILED] 
		\_[REASON] Max value mismatch expected:<4.981786> but was:<0.0>
	Running test: testCalculateExpAndSum     ................  [FAILED] 
		\_[REASON] Sum value mismatch expected:<78781.76> but was:<0.0>
	Running test: testNormalizeSoftmax       ................  [PASS] 
	Running test: testComputeWeightedSum     ................  [PASS] 
	Running test: testSingleHeadFullPipeline ................  [FAILED] 
		\_[REASON] Output mismatch at dimension 0 expected:<0.010636929422616959> but was:<NaN>
	Running test: testMultiHeadAttentionFixed ................  [FAILED] 
		\_[REASON] Output mismatch at head 0, dimension 0 expected:<-0.052507114> but was:<NaN>
Test ran: 8, Failed: 5, Unsupported: 0
```

Both OpenCL and PTX are correct.

**Comment by jjfumero** [comment]
My setup for the SPIR-V:

```bash
$ tornado --devices

Number of Tornado drivers: 2
Driver: SPIR-V
  Total number of SPIR-V devices  : 2
  Tornado device=0:0  (DEFAULT)
	SPIRV -- SPIRV OCL - Intel(R) UHD Graphics 770
		Global Memory Size: 28.8 GB
		Local Memory Size: 64.0 KB
		Workgroup Dimensions: 3
		Total Number of Block Threads: [512]
		Max WorkGroup Configuration: [512, 512, 512]
		Device OpenCL C version: OpenCL C 1.2

  Tornado device=0:1
	SPIRV -- SPIRV LevelZero - Intel(R) UHD Graphics 770
		Global Memory Size: 28.8 GB
		Local Memory Size: 64.0 KB
		Workgroup Dimensions: 3
		Total Number of Block Threads: [512]
		Max WorkGroup Configuration: [512, 512, 512]
		Device OpenCL C version:  (LEVEL ZERO) 1.5
```

**Comment by mikepapadim** [comment]
closed due to a more up to date set of these kernels in #676 @stratika @mairooni
